File: archive/cdktf-ts/Pr824/lib/dns.ts
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { Route53HealthCheck } from '@cdktf/provider-aws/lib/route53-health-check';
import { Route53Record } from '@cdktf/provider-aws/lib/route53-record';
import { Construct } from 'constructs';

export interface DnsProps {
  hostedZoneId: string; // Route53 hosted zone ID (e.g., Z123ABC...)
  recordName: string; // e.g., app.example.com
  primaryAlbDns: string; // DNS name of the primary ALB
  primaryAlbZoneId: string; // Canonical hosted zone ID of the primary ALB
  secondaryAlbDns: string; // DNS name of the secondary ALB
  secondaryAlbZoneId: string; // Canonical hosted zone ID of the secondary ALB
  healthCheckPath?: string; // Optional health check path
  primaryProvider: AwsProvider; // AWS provider for primary region
  secondaryProvider: AwsProvider; // AWS provider for secondary region
  environment: string; // Environment name (e.g., dev, prod)
}

/**
 * Creates:
 * - HTTPS health checks for each region's ALB
 * - Two latency-based alias A-records pointing at each ALB
 */
export class Dns extends Construct {
  constructor(scope: Construct, id: string, props: DnsProps) {
    super(scope, id);

    const path = props.healthCheckPath ?? '/';

    // Health checks (HTTPS) for primary ALB
    const primaryHc = new Route53HealthCheck(this, 'primaryHc', {
      type: 'HTTPS',
      fqdn: props.primaryAlbDns,
      resourcePath: path,
      requestInterval: 30,
      failureThreshold: 3,
      tags: { Name: `${props.environment}-hc-primary` },
      provider: props.primaryProvider,
    });

    // Health checks (HTTPS) for secondary ALB
    const secondaryHc = new Route53HealthCheck(this, 'secondaryHc', {
      type: 'HTTPS',
      fqdn: props.secondaryAlbDns,
      resourcePath: path,
      requestInterval: 30,
      failureThreshold: 3,
      tags: { Name: `${props.environment}-hc-secondary` },
      provider: props.secondaryProvider,
    });

    // Latency alias record (primary)
    new Route53Record(this, 'primaryLatencyRec', {
      zoneId: props.hostedZoneId,
      name: props.recordName,
      type: 'A',
      setIdentifier: 'primary',
      alias: {
        name: props.primaryAlbDns,
        zoneId: props.primaryAlbZoneId,
        evaluateTargetHealth: true,
      },
      latencyRoutingPolicy: {
        region: props.primaryProvider.region!,
      },
      healthCheckId: primaryHc.id,
      provider: props.primaryProvider,
    });

    // Latency alias record (secondary)
    new Route53Record(this, 'secondaryLatencyRec', {
      zoneId: props.hostedZoneId,
      name: props.recordName,
      type: 'A',
      setIdentifier: 'secondary',
      alias: {
        name: props.secondaryAlbDns,
        zoneId: props.secondaryAlbZoneId,
        evaluateTargetHealth: true,
      },
      latencyRoutingPolicy: {
        region: props.secondaryProvider.region!,
      },
      healthCheckId: secondaryHc.id,
      provider: props.secondaryProvider,
    });
  }
}

File: stacks/dns.py
from cdktf import TerraformStack
from cdktf_cdktf_provider_aws.route53_zone import Route53Zone
from cdktf_cdktf_provider_aws.route53_record import Route53Record, Route53RecordAlias
from cdktf_cdktf_provider_aws.route53_health_check import Route53HealthCheck
from constructs import Construct

class DnsModule(Construct):
    def __init__(self, scope: Construct, id: str, primary_provider, compute,
                 environment_suffix: str, migration_phase: str):
        super().__init__(scope, id)

        # Route 53 Hosted Zone
        self.hosted_zone = Route53Zone(self, "hosted-zone",
            provider=primary_provider,
            name=f"payments-{environment_suffix}.example.com",
            tags={
                "Name": f"payment-zone-{environment_suffix}",
                "MigrationPhase": migration_phase
            }
        )

        # Health check for primary ALB
        self.primary_health_check = Route53HealthCheck(self, "primary-health-check",
            provider=primary_provider,
            fqdn=compute.primary_alb.dns_name,
            port=80,
            type="HTTP",
            resource_path="/health",
            failure_threshold=3,
            request_interval=30,
            tags={
                "Name": f"payment-health-primary-{environment_suffix}",
                "MigrationPhase": migration_phase
            }
        )

        # Weighted routing record for primary region
        self.primary_record = Route53Record(self, "primary-record",
            provider=primary_provider,
            zone_id=self.hosted_zone.zone_id,
            name=f"api.payments-{environment_suffix}.example.com",
            type="A",
            set_identifier="primary",
            weighted_routing_policy={"weight": 100},  # ISSUE: Should support variable weights for gradual migration
            health_check_id=self.primary_health_check.id,
            alias=Route53RecordAlias(
                name=compute.primary_alb.dns_name,
                zone_id=compute.primary_alb.zone_id,
                evaluate_target_health=True
            )
        )

        # ISSUE: Missing secondary region DNS record
        # Should have weighted record for secondary region ALB

        # ISSUE: Missing health check for secondary region
